pkg/table/index.go

```diff
@@ -86,15 +80,79 @@ type Index interface {
+	GenIndexKVIter(sc *stmtctx.StatementContext, indexedValue []types.Datum, h kv.Handle, handleRestoreData []types.Datum) IndexKVGenerator
 	// Exist supports check index exists or not.
 	Exist(sc *stmtctx.StatementContext, txn kv.Transaction, indexedValues []types.Datum, h kv.Handle) (bool, kv.Handle, error)
 	// GenIndexKey generates an index key. If the index is a multi-valued index, use GenIndexKVIter instead.
 	GenIndexKey(sc *stmtctx.StatementContext, indexedValues []types.Datum, h kv.Handle, buf []byte) (key []byte, distinct bool, err error)
 	// GenIndexValue generates an index value.
-	GenIndexValue(sc *stmtctx.StatementContext, distinct bool, indexedValues []types.Datum, h kv.Handle, restoredData []types.Datum) ([]byte, error)
+	GenIndexValue(sc *stmtctx.StatementContext, distinct bool, indexedValues []types.Datum, h kv.Handle, restoredData []types.Datum, buf []byte) ([]byte, error)
 	// FetchValues fetched index column values in a row.
 	// Param columns is a reused buffer, if it is not nil, FetchValues will fill the index values in it,
 	// and return the buffer, if it is nil, FetchValues will allocate the buffer instead.
 	FetchValues(row []types.Datum, columns []types.Datum) ([]types.Datum, error)
 }
+
+// IndexKVGenerator generates kv for an index.
+// It could be also used for generating multi-value indexes.
+type IndexKVGenerator struct {
+	index             Index
+	sCtx              *stmtctx.StatementContext
+	handle            kv.Handle
+	handleRestoreData []types.Datum
+
+	// Only used by multi-value index.
+	allIdxVals [][]types.Datum
+	i          int
+	// Only used by non multi-value index.
+	idxVals []types.Datum
+}
+
+// NewIndexKVGenerator creates a new IndexKVGenerator.
+func NewIndexKVGenerator(
+	index Index,
+	stmtCtx *stmtctx.StatementContext,
+	handle kv.Handle,
+	handleRestoredData []types.Datum,
+	mvIndexData [][]types.Datum,
+	singleIdxData []types.Datum,
+) IndexKVGenerator {
+	return IndexKVGenerator{
+		index:             index,
+		sCtx:              stmtCtx,
+		handle:            handle,
+		handleRestoreData: handleRestoredData,
+		allIdxVals:        mvIndexData,
+		i:                 0,
+		idxVals:           singleIdxData,
+	}
+}
+
+// Next returns the next index key and value.
+// For non multi-value indexes, there is only one index kv.
+func (iter *IndexKVGenerator) Next(keyBuf, valBuf []byte) ([]byte, []byte, bool, error) {
+	var val []types.Datum
+	if len(iter.allIdxVals) == 0 {
+		val = iter.idxVals
+	} else {
+		val = iter.allIdxVals[iter.i]
+	}
+	key, distinct, err := iter.index.GenIndexKey(iter.sCtx, val, iter.handle, keyBuf)
+	if err != nil {
+		return nil, nil, false, err
+	}
+	idxVal, err := iter.index.GenIndexValue(iter.sCtx, distinct, val, iter.handle, iter.handleRestoreData, valBuf)
+	if err != nil {
+		return nil, nil, false, err
+	}
+	iter.i++
+	return key, idxVal, distinct, err
+}
+
+// Valid returns true if the generator is exhausted.
```

Comment:
```suggestion
// Valid returns true if the generator is not exhausted.
```

just a guess. In fact I don't know the meaning of `iter.i`